File: MemoryManagement.java
// Filename: MemoryManagement.java
// Author: Unique Karanjit 
// Date: Feb 2, 2025
// Description: A simple example in Java demonstrating garbage collection.

public class MemoryManagement {
    public static void main(String[] args) {
        int[] data = new int[]{1, 2, 3, 4, 5}; // Heap allocation
        processData(data);
        System.out.println("Data in main after function call: " + java.util.Arrays.toString(data));
        // No need to manually free memory—GC will clean up
    }

    static void processData(int[] data) {
        System.out.println("Processing data: " + java.util.Arrays.toString(data));
    }
}

File: MemoryManagement.cpp
// Filename: MemoryManagement.cpp
// Author: Unique Karanjit 
// Date: Feb 2, 2025
// Description: A simple example in C++ demonstrating manual memory management with `new` and `delete`.

#include <iostream>

void processData(int* data, int size) {
    std::cout << "Processing data: ";
    for (int i = 0; i < size; i++) {
        std::cout << data[i] << " ";
    }
    std::cout << std::endl;
}

int main() {
    int* data = new int[5]{1, 2, 3, 4, 5}; // Explicit heap allocation
    processData(data, 5);
    delete[] data; // Manual deallocation to avoid memory leaks
    return 0;
}
